runtime/tsan.c

```diff
@@ -98,7 +98,90 @@
+   Yet we are using it to detect data races in OCaml programs, and even in a
+   mix of OCaml and C code (e.g. between the program and the runtime). To
+   accomplish this, we use a carefully chosen mapping of OCaml memory accesses
+   to C11 accesses.
+
+   For each type of OCaml memory access on the left, we signal to TSan the
+   operations on the right:
+
+   OCaml access      | C11 equivalent          | TSan view
+   ------------------|-------------------------|--------------------------------
+   Atomic load       | fence(acquire)          | __tsan_atomic64_load(seq_cst)
+                     | atomic_load(seq_cst)    |
+   ------------------|-------------------------|--------------------------------
+   Atomic store      | fence(acquire)          | __tsan_atomic_exchange(seq_cst)
+                     | atomic_exchange(seq_cst)|
+                     | fence(release)          |
+   ------------------|-------------------------|--------------------------------
+   Non-atomic load   | atomic_load(relaxed)    | __tsan_read8()
+   ------------------|-------------------------|--------------------------------
+   Non-atomic store  | fence(acquire)          | __tsan_write8()
+   (assignment, int) | atomic_store(release)   |
+   ------------------|-------------------------|--------------------------------
+   Non-atomic store  | fence(acquire)          | __tsan_write8()
+   (assignment, ptr) | atomic_store(release)   |
+                     | fence(release)          |
+   ------------------|-------------------------|--------------------------------
+   Non-atomic store  | plain store             | __tsan_writeN()
+   (unaligned size)  |                         |
+   ------------------|-------------------------|--------------------------------
+
+   3.1. False negatives
+
+   There should be no false negatives, i.e., all data races (in the OCaml
+   sense) on visited code paths should be detected (modulo TSan limitations
+   such as the finite history of memory accesses).
+
+   3.2. False positives
+
+   The mapping that we use should not incur any false positives in pure OCaml
+   code, i.e., all data races reported between two accesses from OCaml are
+   true data races (in the OCaml sense of the term).
+
+   In mixed C-OCaml code, rare false positives may occur in the following cases:
+
+   - A value is initialized from C without using `caml_initialize` (allowed by
+     the FFI rules on the condition that the GC does not run between the
+     allocation and the end of initialization) and a conflicting access is made
+     from OCaml after publication to other threads. There is no data race due
+     to data dependencies, but TSan does not take data dependencies into
+     account.
+   - A field is accessed from C with `Field`, or more generally using a
+     `volatile value *` or a relaxed atomic access, and that field is modified
+     concurrently by OCaml code. Because caml_modify as a plain write for
```

Comment:
is a plain write?

Reply:
This sentence is not so easy to understand for me. Maybe you could mention (with slightly more details) the translation / TSan effect of `caml_modify` and `caml_initialize` somewhere above, for example right before section 3.1?

Reply:
> is a plain write?

This was a typo, thanks for spotting it. The right phrasing is: is instrumented as a plain write.

> This sentence is not so easy to understand for me. Maybe you could mention (with slightly more details) the translation / TSan effect of caml_modify and caml_initialize somewhere above, for example right before section 3.1?

I added a paragraph about this. There was also a missing row in the table that I added.